File: Ex0408_StringToInteger/StringToInteger.c
#include <stdio.h>

int my_atoi(const char s[], size_t length)
{
	int rslt = 0;
	int sign = 1;
	int i = 0;
	while (s[i] == '-')
	{
		sign *= -1;
		i++;
	}
	while (i < (int)length && s[i] >= '0' && s[i] <= '9')
	{
		rslt *= 10;
		rslt += s[i] - '0';
		i++;
	}
	return rslt * sign;
}

int main(void)
{
	char s[] = "-1231904";
	printf("Origin string: %s\n", s);
	int num = my_atoi(s, sizeof(s)/sizeof(s[0]));
	printf("Translated integer: %d\n", num);
	return 0;
}

File: Ex0408_StringToInteger/StringToInteger.java

public class StringToInteger {
    public static void main(String[] args) {
        String s = "-1231904";
        System.out.println("Origin string: " + s);
        int num = myAtoi(s);
        System.out.println("Translated integer: " + num);
    }

    public static int myAtoi(String s) {
        if (s == null) {
            throw new NullPointerException();
        }
        int rslt = 0;
        int sign = 1;
        int i = 0;
        while (s.charAt(i) == '-') {
            sign *= -1;
            i++;
        }
        while (i < s.length() && s.charAt(i) >= '0'
                && s.charAt(i) <= '9') {
            rslt *= 10;
            rslt += s.charAt(i) - '0';
            i++;
        }
        return rslt * sign;
    }
}
